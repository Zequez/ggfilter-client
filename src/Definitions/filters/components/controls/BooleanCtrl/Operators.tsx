import * as React from 'react';
import * as th from './BooleanCtrl.sass';
import * as cx from 'classnames';

import TooltipLabel from './TooltipLabel'

export type Mode = 'and' | 'or' | 'xor';
type OperatorsProps = {
  modes: Mode[];
  value: Mode;
  onChange: (mode: Mode) => void;
};

export default class Operators extends React.Component<OperatorsProps> {
  onClick = () => {
    let { value, modes } = this.props;
    let newMode = modes[modes.indexOf(value) + 1] || modes[0];
    this.props.onChange && this.props.onChange(newMode);
  }

  render () {
    let { modes, value } = this.props;

    let modesTitles = modes.map((m) => m.toUpperCase());
    let tooltip = modesTitles.join('/');

    let className = cx(
      th.BooleanCtrl__Operators,
      th['BooleanCtrl__Operators_' + value]
    );

    let valueIndex = modes.indexOf(value);
    let transform = {
      transform: `translateY(-${100 * valueIndex}%)`
    };

    return (
      <TooltipLabel
        className={className}
        tooltip={tooltip}
        onClick={this.onClick}>
        {modes.map((mode, i) =>
          <span
            key={mode}
            style={transform}
            className={cx(th.BooleanCtrl__Operator, {
              [th.BooleanCtrl__Operator_active]: mode === value
            })}>
            {modesTitles[i]}
          </span>
        )}
      </TooltipLabel>
    )
  }
}
