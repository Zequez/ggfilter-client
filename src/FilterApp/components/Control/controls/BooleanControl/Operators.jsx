import th from './BooleanControl.sass'
import React, { PropTypes as t, Component } from 'react'
import cx from 'classnames'

import TooltipLabel from './TooltipLabel'

export default class Operators extends Component {
  static propTypes = {
    modes: t.arrayOf(t.oneOf(['and', 'or', 'xor'])).isRequired,
    value: t.oneOf(['and', 'or', 'xor']).isRequired,
    onChange: t.func.isRequired
  }

  onClick = () => {
    let { value, modes } = this.props
    let newMode = modes[modes.indexOf(value) + 1] || modes[0]
    this.props.onChange && this.props.onChange(newMode)
  }

  render () {
    let { modes, value } = this.props

    let modesTitles = modes.map((m) => m.toUpperCase())
    let tooltip = modesTitles.join('/')

    let className = cx(
      th.BooleanControl__Operators,
      th['BooleanControl__Operators_' + value]
    )

    let valueIndex = modes.indexOf(value)
    let transform = {
      transform: `translateY(-${100 * valueIndex}%)`
    }

    return (
      <TooltipLabel
        className={className}
        tooltip={tooltip}
        onClick={this.onClick}>
        {modes.map((mode, i) =>
          <span
            key={mode}
            style={transform}
            className={cx(th.BooleanControl__Operator, {
              [th.BooleanControl__Operator_active]: mode === value
            })}>
            {modesTitles[i]}
          </span>
        )}
      </TooltipLabel>
    )
  }
}
