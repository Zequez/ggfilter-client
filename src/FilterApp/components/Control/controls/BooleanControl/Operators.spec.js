import React from 'react'
import renderer from 'react-test-renderer'
import Operators from './Operators'

describe('Operators', () => {
  it('returns change whenever you click it', () => {
    let receiveChange = jest.fn()
    const component = renderer.create(<Operators
      modes={['and', 'or', 'xor']}
      value='or'
      onChange={receiveChange}
    />)
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()

    expect(tree.props.className).toMatch(/Operators_or/i)

    tree.props.onClick()
    expect(receiveChange).toHaveBeenCalledWith('xor')
  })
})
