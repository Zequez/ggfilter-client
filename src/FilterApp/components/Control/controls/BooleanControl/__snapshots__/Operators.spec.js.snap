exports[`Operators returns change whenever you click it 1`] = `
<label
  className="BooleanControl__Operators BooleanControl__Operators_or"
  disabled={false}
  onClick={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onTouchStart={[Function]}>
  <span
    className="BooleanControl__Operator">
    AND
  </span>
  <span
    className="BooleanControl__Operator BooleanControl__Operator_active">
    OR
  </span>
  <span
    className="BooleanControl__Operator">
    XOR
  </span>
</label>
`;
